src/deviceinput.cc

```diff
@@ -625,9 +660,10 @@ void InputMapper::Impl::sequenceTimeout()
     {
       execAction(m_lastState.second->action, DeviceKeyMap::Result::PartialHit);
     }
-    else if (m_vdev && !m_events.empty())
+    else if (m_vmouse && m_vkeyboard && !m_events.empty())
```

Comment:
consider helper method e.g. `hasVirtualDevices()` instead of `m_vmouse && m_vkeyboard`